isogeny_primes.py

```diff
@@ -730,85 +781,140 @@ def get_C_primes(K, G_K, frak_q, epsilons, q_class_group_order, loop_curves=Fals
+                        logging.debug("Emergency aux prime added: {}".format(candidate))
+                    list_of_gens.remove(emergency_gen)
+            i += 1
+
+        if list_of_gens:
+            raise RuntimeError("We have been unable to add enough emergency "
+                            "auxiliary primes. Try increasing the `B` parameter above.")
+    else:
+        a_good_prime = completely_split_rat_primes[0]
+        candidate = K.primes_above(a_good_prime)[0]
+        if a_good_prime > norm_bound:
+            aux_primes.append(candidate)
+            logging.debug("Emergency aux prime added: {}".format(candidate))
+
+    embeddings = K.embeddings(Kgal)
+    if K.is_galois():
+        G_K = K.galois_group()
+        epsilons = get_pre_type_one_two_epsilons(K.degree(), galgp=G_K)
+    else:
+        epsilons = get_pre_type_one_two_epsilons(K.degree())
+
+    # Do the unit computation first
+
+    unit_gens = K.unit_group().gens_values()
+
+    divs_from_units = {eps : gcd([ (eps_exp(u, eps, embeddings) - 1).absolute_norm() for u in unit_gens]) for eps in epsilons}
 
-    try:
-        it = Kgal.primes_of_degree_one_iter(max_iterations=1000)
-        aux_prime_count = 2
-        while aux_prime_count > 0:
-            aux_prime_candidate = next(it)
-            if (not contains_imaginary_quadratic) or (not aux_prime_candidate.is_principal()):
-                if aux_prime_candidate.norm() > norm_bound:
-                    aux_primes.append(aux_prime_candidate)
-                    logging.info("Emergency aux prime added")
-                aux_prime_count -= 1
-    except StopIteration:
-        # This should never happen, because in this case we are probably containing
-        # the Hilbert class field of an imaginary quadratic field and therefore
-        # would already have raised above, but I have been unable to prove this.
-        raise ValueError("The galois closure of your number field contains an "
-                "imaginary quadratic field, and we are unable to find a "
-                "non-principal auxiliary prime which is necessary for the algorithm "
-                "to work.")
-
-    G_K = Kgal.galois_group()
-    C_Kgal = Kgal.class_group()
-    logging.debug("C_Kgal = {}".format(C_Kgal))
-    epsilons = get_pre_type_one_two_epsilons(G_K)
     tracking_dict = {}
+    frob_polys_dict = {}
 
     for q in aux_primes:
-        q_class_group_order = C_Kgal(q).multiplicative_order()
+        q_class_group_order = C_K(q).multiplicative_order()
         # these will be dicts with keys the epsilons, values sets of primes
-        AB_primes_dict = get_AB_primes(G_K,q,epsilons, q_class_group_order)
-        C_primes_dict = get_C_primes(Kgal,G_K, q, epsilons, q_class_group_order, loop_curves)
+        AB_integers_dict = get_AB_integers(embeddings,q,epsilons, q_class_group_order)
+        C_integers_dict, frob_polys = get_C_integers(Kgal,embeddings, q, epsilons, q_class_group_order, loop_curves)
```

Comment:
I think it makes no sense to get the `frob_polys` from the `get_C_integers` function. The `get_weil_polys` function is ridiculously fast so it is something that we can just compute again on the fly. The bottleneck will be in other parts of the code. Especially since this ties the `get_C_integers` functionality together with the PIL code in an unatural way and make the code less maintainable.

Reply:
If you really want the `frob_polys_dict` (for caching and efficiency reasons) then my suggestion is that you compute it outside of the `get_C_integers` function using a seperate function whose sole purpose is to compute this dict. And give it as an (optional?) input to `get_C_integers`.

Reply:
Good point -- the next commit has gone for the second of these suggestions, to compute the polys separately and pass the list to `get_C_integers`.